### PR TITLE
[sui framework][verifier] Make Transfer functions restricted

### PR DESCRIPTION
**Pull request by tnowacki**
- Transfer functions are now private to the module that defines the type
- This restriction is lifted for types with store

- delete_child_object is the odd function out, and is not "private". The assumption is that if you were able to get the VersionedID, the module has already "destroyed" the object, and this function is just for cleanup.

- I will need help with being pointed to what docs to fix

### CODE REVIEW

**Comment by emmazzz** [comment]
I want to understand better the context of this change. Is it so that modules can define arbitrary condition under which the transfer can happen, such as "you can only transfer your coin if your address ends with 0 "? And why is the restriction lifted for objects with store ability?

**Comment by tnowacki** [comment]
> Is it so that modules can define arbitrary condition under which the transfer can happen

Yes

> And why is the restriction lifted for objects with store ability?

It is an ergonomics win, without changing expressiveness. If the type has key+store, anyone can wrap it into a simple object that exposes all of the transfer operations, bypassing any logic in the defining module. 
In other words, if you have store, you cannot actually restrict any transfers, so the transfer functions are no longer actually private.
We believe that more often than not, this is actually what you want. So relaxing it for store improves the ergonomics in the common case, and removes the dance of going into/out of a wrapper with exposed transfer functions

**Review by emmazzz — APPROVED** [review]
lgtm

File: crates/sui-framework/sources/Coin.move [hunk 1]
```diff
@@ -9,7 +9,7 @@ module Sui::Coin {
     use Std::Vector;
 
     /// A coin of type `T` worth `value`. Transferable but not storable
```
Comment (emmazzz) [inline crates/sui-framework/sources/Coin.move:11]: This comment should probably be changed now that coin is storable.

File: crates/sui-framework/sources/Bag.move [hunk 2]
```diff
@@ -118,16 +118,28 @@ module Sui::Bag {
...
     /// Transfer the entire Bag to `recipient`.
-    public(script) fun transfer(c: Bag, recipient: address, _ctx: &mut TxContext) {
+    public(script) fun transfer_(c: Bag, recipient: address) {
         Transfer::transfer(c, recipient)
     }
 
+    /// Transfer the entire Bag to `recipient`.
+    public fun transfer(c: Bag, recipient: address) {
+        Transfer::transfer(c, recipient)
+    }
```
Comment (lxfind) [inline crates/sui-framework/sources/Bag.move:134]: It took me a while to realize the difference between these two functions (one is entry, and one is not). The difference is very subtle. Worth with either an explicitly comment or some better naming convention to highlight the difference.

Comment (tnowacki) [reply crates/sui-framework/sources/Bag.move:134]: I can do that, but I also want to fix this rather quickly (probably early next week). Where entry != visibility.

**Review by lxfind — APPROVED** [review]

**Pull request merged**